lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,961 @@
+        get_socket,      /* get underlying socket */
+        NULL,            /* get debug state */
+        {{{NULL}}}};
+
+    mrb_state *mrb = ctx->shared->mrb;
+    int gc_arena = mrb_gc_arena_save(mrb);
+    mrb_gc_protect(mrb, env);
+
+    /* create subreq */
+    struct st_mruby_subreq_t *subreq = h2o_mem_alloc(sizeof(*subreq));
+    h2o_req_t *super = &subreq->super;
+    memset(&subreq->conn, 0, sizeof(subreq->conn));
+    subreq->ctx = ctx;
+    subreq->receiver = mrb_nil_value();
+    subreq->refs.request = mrb_nil_value();
+    subreq->refs.input_stream = mrb_nil_value();
+    h2o_buffer_init(&subreq->buf, &h2o_socket_buffer_prototype);
+    subreq->shortcut.response = NULL;
+    subreq->shortcut.body = NULL;
+    subreq->state = INITIAL;
+    subreq->chain_proceed = 0;
+
+    /* initialize super and conn */
+    subreq->conn.super.ctx = ctx->shared->ctx;
+    h2o_init_request(&subreq->super, &subreq->conn.super, NULL);
+    super->is_subrequest = 1;
+    h2o_ostream_t *ostream = h2o_add_ostream(super, H2O_ALIGNOF(*ostream), sizeof(*ostream), &super->_ostr_top);
+    ostream->do_send = subreq_ostream_send;
+    subreq->conn.super.hosts = ctx->handler->pathconf->global->hosts;
+    subreq->conn.super.connected_at = (struct timeval){0}; /* no need because subreq won't logged */
+    subreq->conn.super.id = 0; /* currently conn->id is used only for logging, so set zero as a meaningless value */
+    subreq->conn.super.callbacks = &callbacks;
+
+
+    /* retrieve env variables */
+    mrb_value scheme = mrb_nil_value();
+    mrb_value method = mrb_nil_value();
+    mrb_value script_name = mrb_nil_value();
+    mrb_value path_info = mrb_nil_value();
+    mrb_value query_string = mrb_nil_value();
+    mrb_value rack_input = mrb_nil_value();
+    mrb_value http_host = mrb_nil_value();
+    mrb_value server_name = mrb_nil_value();
+    mrb_value server_port = mrb_nil_value();
+    mrb_value server_addr = mrb_nil_value();
+    mrb_value remote_addr = mrb_nil_value();
+    mrb_value remote_port = mrb_nil_value();
+    mrb_value server_protocol = mrb_nil_value();
+    mrb_value remaining_delegations = mrb_nil_value();
+    mrb_value remaining_reprocesses = mrb_nil_value();
+    mrb_value rack_errors = mrb_nil_value();
+
+#define RETRIEVE_ENV(val, stringify) do { \
+    val = value; \
+    if (!mrb_nil_p(val) && stringify) { \
+        val = h2o_mruby_to_str(mrb, val); \
+        if (mrb->exc != NULL) \
+            goto Failed; \
+    } \
+} while (0)
+#define COND0(str, lit, pos) (sizeof(lit) - 1 <= (pos) || (str)[pos] == (lit)[pos])
+#define COND1(str, lit, pos) (COND0(str, lit, pos) && COND0(str, lit, pos + 1) && COND0(str, lit, pos + 2))
+#define COND2(str, lit, pos) (COND1(str, lit, pos) && COND1(str, lit, pos + 3) && COND1(str, lit, pos + 6))
+#define COND(str, lit)       (COND2(str, lit, 0)   && COND2(str, lit, 9)       && COND2(str, lit, 18))
+#define CHECK_KEY(lit) ((sizeof(lit) - 1) == keystr_len && COND(keystr, lit))
```

Comment:
May I ask the reason for not using `memcmp` here?

I would assume that the C compiler will generate efficient code to compare a prefixed number of bytes. For example, for a CPU that does not have penalty on unaligned load (e.g. x86-64 without cache miss), a compiler might generate 64-bit compare instruction to compare 8 bytes at once.

Reply:
This expects compiler optimizations. If we use `memcmp`, comparing a hash key with known keys (e.g. `SCRIPT_NAME`) take the cost in O(nm) where n is the lengthes of hash key and m is sum of the length of known keys. OTOH if we directly write character comparisons like my commit, `-O2` compilation optimizes it by omitting needless comparisons just like using prefix tree, and reduces the cost to O(n). I checked the behaviors of gcc and clang, and took some benchmarks that showed this way is certainly faster than `memcmp`s and other methods (like calculating hashes, etc)